onnxruntime/core/providers/coreml/builders/impl/flatten_op_builder.cc

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/coreml/builders/helper.h"
+
+#ifdef __APPLE__
+#include "core/providers/coreml/builders/model_builder.h"
+#endif
+#include "core/providers/coreml/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class FlattenOpBuilder : public BaseOpBuilder {
+  // Add operator related
+#ifdef __APPLE__
+ private:
+  [[nodiscard]] Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                             const logging::Logger& logger) const override;
+#endif
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+};
+
+// Add operator related
+
+#ifdef __APPLE__
+
+Status FlattenOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                               const Node& node,
+                                               const logging::Logger& logger) const {
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(model_builder, node);
+
+  // Note: ONNX Flatten corresponds to CoreML FlattenTo2DLayerParams
+  auto* coreml_flatten = layer->mutable_flattento2d();
+
+  NodeAttrHelper helper(node);
+  const int64_t axis = helper.Get("axis ", 1);
+  coreml_flatten->set_axis(axis);
+
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(std::move(layer));
+
+  return Status::OK();
+}
+#endif
+
+// Operator support related
+
+bool FlattenOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& /*input_params*/,
+                                         const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  if (input_shape.empty()) {
+    LOGS(logger, VERBOSE) << "Flatten does not support empty input shape";
+    return false;
+  }
+
+  const auto input_rank = input_shape.size();
+  if (input_rank < 3) {
+    LOGS(logger, VERBOSE) << "Flatten only supports input rank greater than equal to 3, input rank is"
+                          << input_rank;
+    return false;
```

Comment:
Is this a requirement for flatten? didn't notice it in the spec and the example show it working with input that has a rank of 1.

Reply:
good catch. the requirement comes from FlattenLayerParams - was originally looking at that one in CoreML. For FlattenTo2D, it does not require the input rank >= 3.